src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -346,6 +379,9 @@ Node* IdealLoopTree::reassociate_add_sub(Node* n1, int inv1_idx, int inv2_idx, P
       inv = new AddINode(n_inv1, inv2);
     }
     phase->register_new_node(inv, phase->get_early_ctrl(inv));
+    if (n1_is_cmp) {
```

Comment:
CmpNode is subclass of SubNode. if n is CmpI/L, n->is_Sub() is also true. 
can we use the old logic for your new comparison expressions? 

yes, we still need to check if n1 is CmpNode or SubNode here.

Reply:
I think CmpNode has the same re-association rule as SubNode. 
take your own example, 
>  inv1 == (x - inv2) => ( inv1 + inv2 ) == x

Cmp(inv1, (x-inv2)) => Eq(0, Sub(inv1, (x-inv2)) => Eq(0, Sub(inv1+inv2, x)) => Cmp(inv1+inv2, x)

Originally, n1->is_Sub() covers both CmpNode and SubNode. I don't you need to split them it into 2 cases. 

I think you only to check if n1 Cmp or Sub when you are going to return Cmp/SubNode

Reply:
The rules are not always the same. For example

> inv1 - (inv2 - x)     => (inv1 - inv2) + x
> inv1  == (inv2 - x) => (-inv1 + inv2) == x

The signs of inv1 and inv2 are flipped. I don't think we can fold these into a single conditional

Reply:
okay. I see you mean.